File: src/llm/ollama.ts
import { ChatMessage } from "../types/chat";
import { ModelConfig } from "../types/config";
import { logger } from "../utils/logger";

export interface OllamaChatRequest {
    model: string;
    messages: ChatMessage[];
    temperature?: number;
    max_tokens?: number;
    stream?: boolean;
}

export interface OllamaChatResponse {
    message: {
        role: 'assistant';
        content: string;
    };
    done: boolean;
}

/**
 * Send a chat request to a locally running Ollama instance.
 */
export async function ollamaChat(
    config: ModelConfig,
    messages: ChatMessage[],
    options?: {
        stream?: boolean;
    }
): Promise<string> {
    const request: OllamaChatRequest = {
        model: config.name,
        messages,
        temperature: config.temperature,
        max_tokens: config.maxTokens,
        stream: options?.stream ?? false,
    };

    logger.debug('ollama.chat.request', {
        model: config.name,
        messages: messages.length,
        temperature: config.temperature,
        maxTokens: config.maxTokens,
        stream: request.stream,
    });

    const started = Date.now();

    const response = await fetch(
        'http://localhost:11434/api/chat',
        {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
            },
            body: JSON.stringify(request),
        }
    );

    const latencyMs = Date.now() - started;

    if (!response.ok) {
        logger.error('ollama.chat.http_error', {
            status: response.status,
            model: config.name,
            latencyMs,
        });

        throw new Error(
            `Ollama chat request failed (${response.status})`
        );
    }

    if (request.stream) {
        logger.error('ollama.chat.streaming_not_implemented', {
            model: config.name,
        });
        throw new Error('Streaming not implemented yet');
    }

    const data = (await response.json()) as OllamaChatResponse;

    if (!data?.message?.content) {
        logger.error('ollama.chat.invalid_response', {
            model: config.name,
            latencyMs,
        });
        throw new Error('Invalid response from Ollama');
    }

    logger.info('ollama.chat.success', {
        model: config.name,
        latencyMs,
    });

    return data.message.content;
}

File: rust/rlm/src/llm/ollama.rs
use reqwest::Client;
use serde::{Deserialize, Serialize};

use crate::types::chat::ChatMessage;
use crate::types::config::ModelConfig;
use crate::utils::logger;

#[derive(Debug, Serialize)]
struct OllamaChatRequest<'a> {
    model: &'a str,
    messages: &'a [ChatMessage],
    #[serde(skip_serializing_if = "Option::is_none")]
    temperature: Option<f32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(rename = "max_tokens")]
    max_tokens: Option<u32>,
    #[serde(default)]
    stream: bool,
}

#[derive(Debug, Deserialize)]
struct OllamaChatResponseMessage {
    role: String,
    content: String,
}

#[derive(Debug, Deserialize)]
struct OllamaChatResponse {
    message: OllamaChatResponseMessage,
    done: bool,
}

pub async fn ollama_chat(
    client: &Client,
    config: &ModelConfig,
    messages: &[ChatMessage],
    stream: bool,
) -> anyhow::Result<String> {
    let request = OllamaChatRequest {
        model: &config.name,
        messages,
        temperature: config.temperature,
        max_tokens: config.max_tokens,
        stream,
    };

    logger::debug(
        "ollama.chat.request",
        Some(serde_json::json!({
            "model": config.name,
            "messages": messages.len(),
            "temperature": config.temperature,
            "maxTokens": config.max_tokens,
            "stream": stream,
        })),
    );

    let started = std::time::Instant::now();

    let response = client
        .post("http://localhost:11434/api/chat")
        .json(&request)
        .send()
        .await?;

    let latency_ms = started.elapsed().as_millis();

    if !response.status().is_success() {
        logger::error(
            "ollama.chat.http_error",
            Some(serde_json::json!({
                "status": response.status().as_u16(),
                "model": config.name,
                "latencyMs": latency_ms,
            })),
        );
        anyhow::bail!(
            "Ollama chat request failed ({})",
            response.status().as_u16()
        );
    }

    if stream {
        logger::error(
            "ollama.chat.streaming_not_implemented",
            Some(serde_json::json!({ "model": config.name })),
        );
        anyhow::bail!("Streaming not implemented yet");
    }

    let data: OllamaChatResponse = response.json().await?;

    if data.message.content.is_empty() {
        logger::error(
            "ollama.chat.invalid_response",
            Some(serde_json::json!({
                "model": config.name,
                "latencyMs": latency_ms,
            })),
        );
        anyhow::bail!("Invalid response from Ollama");
    }

    logger::info(
        "ollama.chat.success",
        Some(serde_json::json!({
            "model": config.name,
            "latencyMs": latency_ms,
        })),
    );

    Ok(data.message.content)
}
